File: test_setup.php
<?php
/**
 * Contact Form Setup Test
 * 
 * This file helps you test your server configuration and email setup.
 * Run this file in your browser to check if everything is working correctly.
 */

// Include configuration
require_once 'config.php';

// Set content type
header('Content-Type: text/html; charset=UTF-8');
?>
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Contact Form Setup Test</title>
    <style>
        body { font-family: Arial, sans-serif; max-width: 800px; margin: 0 auto; padding: 20px; }
        .test-section { margin: 20px 0; padding: 15px; border: 1px solid #ddd; border-radius: 5px; }
        .success { background-color: #d4edda; border-color: #c3e6cb; color: #155724; }
        .warning { background-color: #fff3cd; border-color: #ffeaa7; color: #856404; }
        .error { background-color: #f8d7da; border-color: #f5c6cb; color: #721c24; }
        .info { background-color: #d1ecf1; border-color: #bee5eb; color: #0c5460; }
        h1 { color: #333; }
        h2 { color: #555; margin-top: 0; }
        code { background: #f4f4f4; padding: 2px 4px; border-radius: 3px; }
        .config-value { font-weight: bold; }
    </style>
</head>
<body>
    <h1>Contact Form Setup Test</h1>
    <p>This page tests your contact form setup and server configuration.</p>

    <?php
    // Test 1: PHP Version
    echo '<div class="test-section ' . (version_compare(PHP_VERSION, '5.6.0', '>=') ? 'success' : 'error') . '">';
    echo '<h2>PHP Version Test</h2>';
    echo '<p>Current PHP Version: <span class="config-value">' . PHP_VERSION . '</span></p>';
    if (version_compare(PHP_VERSION, '5.6.0', '>=')) {
        echo '<p>✓ PHP version is compatible</p>';
    } else {
        echo '<p>✗ PHP version is too old. Please upgrade to PHP 5.6 or higher.</p>';
    }
    echo '</div>';

    // Test 2: Configuration
    echo '<div class="test-section info">';
    echo '<h2>Configuration Test</h2>';
    echo '<p>Recipient Email: <span class="config-value">' . RECIPIENT_EMAIL . '</span></p>';
    echo '<p>Sender Email: <span class="config-value">' . SENDER_EMAIL . '</span></p>';
    echo '<p>Website Name: <span class="config-value">' . WEBSITE_NAME . '</span></p>';
    echo '<p>Debug Mode: <span class="config-value">' . (DEBUG_MODE ? 'Enabled' : 'Disabled') . '</span></p>';
    echo '<p>Logging: <span class="config-value">' . (ENABLE_LOGGING ? 'Enabled' : 'Disabled') . '</span></p>';
    
    $configErrors = validateConfig();
    if (empty($configErrors)) {
        echo '<p>✓ Configuration appears to be valid</p>';
    } else {
        echo '<p>⚠ Configuration issues found:</p><ul>';
        foreach ($configErrors as $error) {
            echo '<li>' . htmlspecialchars($error) . '</li>';
        }
        echo '</ul>';
    }
    echo '</div>';

    // Test 3: Mail Function
    echo '<div class="test-section ' . (function_exists('mail') ? 'success' : 'error') . '">';
    echo '<h2>Mail Function Test</h2>';
    if (function_exists('mail')) {
        echo '<p>✓ PHP mail() function is available</p>';
        
        // Test sending a simple email
        if (isset($_GET['test_email']) && $_GET['test_email'] === '1') {
            $testSubject = 'Contact Form Test Email';
            $testMessage = 'This is a test email from your contact form setup. If you receive this, your email configuration is working!';
            $testHeaders = 'From: ' . SENDER_EMAIL . "\r\n" . 'Content-Type: text/plain; charset=UTF-8';
            
            if (mail(RECIPIENT_EMAIL, $testSubject, $testMessage, $testHeaders)) {
                echo '<p>✓ Test email sent successfully to ' . RECIPIENT_EMAIL . '</p>';
            } else {
                echo '<p>✗ Failed to send test email. Check your mail server configuration.</p>';
            }
        } else {
            echo '<p><a href="?test_email=1">Click here to send a test email</a></p>';
        }
    } else {
        echo '<p>✗ PHP mail() function is not available. Check your server configuration.</p>';
    }
    echo '</div>';

    // Test 4: File Permissions
    echo '<div class="test-section">';
    echo '<h2>File Permissions Test</h2>';
    
    $writableDir = is_writable(dirname(__FILE__));
    echo '<div class="' . ($writableDir ? 'success' : 'warning') . '">';
    if ($writableDir) {
        echo '<p>✓ Current directory is writable (needed for logging)</p>';
    } else {
        echo '<p>⚠ Current directory is not writable. Logging may not work.</p>';
    }
    echo '</div>';
    
    // Test log file creation
    if (ENABLE_LOGGING) {
        $logFile = LOG_FILE;
        $canCreateLog = @file_put_contents($logFile, 'Test log entry - ' . date('Y-m-d H:i:s') . "\n", FILE_APPEND | LOCK_EX);
        echo '<div class="' . ($canCreateLog ? 'success' : 'error') . '">';
        if ($canCreateLog) {
            echo '<p>✓ Log file can be created/written to</p>';
        } else {
            echo '<p>✗ Cannot write to log file. Check directory permissions.</p>';
        }
        echo '</div>';
    }
    echo '</div>';

    // Test 5: Required Extensions
    echo '<div class="test-section">';
    echo '<h2>PHP Extensions Test</h2>';
    
    $requiredExtensions = ['filter', 'json'];
    $allExtensionsLoaded = true;
    
    foreach ($requiredExtensions as $ext) {
        $loaded = extension_loaded($ext);
        $allExtensionsLoaded = $allExtensionsLoaded && $loaded;
        echo '<p>' . ($loaded ? '✓' : '✗') . ' ' . $ext . ' extension: ' . ($loaded ? 'Loaded' : 'Not loaded') . '</p>';
    }
    
    echo '<div class="' . ($allExtensionsLoaded ? 'success' : 'error') . '">';
    if ($allExtensionsLoaded) {
        echo '<p>All required PHP extensions are loaded</p>';
    } else {
        echo '<p>Some required PHP extensions are missing</p>';
    }
    echo '</div>';
    echo '</div>';

    // Test 6: Form Files
    echo '<div class="test-section">';
    echo '<h2>Form Files Test</h2>';
    
    $requiredFiles = ['contact.html', 'send_email.php', 'css/contact.css', 'config.php'];
    $allFilesExist = true;
    
    foreach ($requiredFiles as $file) {
        $exists = file_exists($file);
        $allFilesExist = $allFilesExist && $exists;
        echo '<p>' . ($exists ? '✓' : '✗') . ' ' . $file . ': ' . ($exists ? 'Found' : 'Missing') . '</p>';
    }
    
    echo '<div class="' . ($allFilesExist ? 'success' : 'error') . '">';
    if ($allFilesExist) {
        echo '<p>All required files are present</p>';
    } else {
        echo '<p>Some required files are missing</p>';
    }
    echo '</div>';
    echo '</div>';

    // Summary
    echo '<div class="test-section info">';
    echo '<h2>Next Steps</h2>';
    echo '<ol>';
    echo '<li>Update <code>config.php</code> with your actual email addresses</li>';
    echo '<li>Test the contact form by opening <code>contact.html</code></li>';
    echo '<li>Check your email for test submissions</li>';
    echo '<li>Customize the styling in <code>css/contact.css</code> if needed</li>';
    echo '<li>Deploy to your production server</li>';
    echo '</ol>';
    echo '<p><strong>Security Note:</strong> Remove or restrict access to this test file in production!</p>';
    echo '</div>';
    ?>

    <div class="test-section info">
        <h2>Troubleshooting</h2>
        <ul>
            <li>If emails aren't being sent, check your server's mail configuration</li>
            <li>Check PHP error logs for detailed error messages</li>
            <li>Ensure your hosting provider allows email sending</li>
            <li>Consider using SMTP instead of the mail() function for better reliability</li>
            <li>Test with different email providers (some may block certain servers)</li>
        </ul>
    </div>
</body>
</html>

File: test_setup.py
#!/usr/bin/env python3
"""
Test script to verify Gmail contact form setup
"""

import os
import sys
from dotenv import load_dotenv
import re

def test_dependencies():
    """Test if required Python packages are installed"""
    print("🔍 Testing Python dependencies...")
    
    try:
        import flask
        print(f"✅ Flask {flask.__version__} installed")
    except ImportError:
        print("❌ Flask not installed. Run: pip install flask")
        return False
    
    try:
        import dotenv
        print(f"✅ python-dotenv installed")
    except ImportError:
        print("❌ python-dotenv not installed. Run: pip install python-dotenv")
        return False
    
    return True

def test_environment():
    """Test environment variables configuration"""
    print("\n🔍 Testing environment configuration...")
    
    # Load environment variables
    load_dotenv()
    
    gmail_address = os.getenv('GMAIL_ADDRESS')
    gmail_password = os.getenv('GMAIL_APP_PASSWORD')
    recipient_email = os.getenv('RECIPIENT_EMAIL')
    
    # Check Gmail address
    if not gmail_address:
        print("❌ GMAIL_ADDRESS not set in .env file")
        return False
    elif not re.match(r'^[a-zA-Z0-9._%+-]+@gmail\.com$', gmail_address):
        print(f"❌ GMAIL_ADDRESS '{gmail_address}' is not a valid Gmail address")
        return False
    else:
        print(f"✅ GMAIL_ADDRESS: {gmail_address}")
    
    # Check Gmail app password
    if not gmail_password:
        print("❌ GMAIL_APP_PASSWORD not set in .env file")
        print("📖 See SETUP_GUIDE.md for instructions on creating an App Password")
        return False
    elif len(gmail_password.replace(' ', '')) != 16:
        print(f"❌ GMAIL_APP_PASSWORD should be 16 characters (currently {len(gmail_password)} characters)")
        print("📖 Make sure you're using a Gmail App Password, not your regular password")
        return False
    else:
        print(f"✅ GMAIL_APP_PASSWORD: {'*' * len(gmail_password)} (length: {len(gmail_password)})")
    
    # Check recipient email
    if recipient_email:
        if not re.match(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$', recipient_email):
            print(f"❌ RECIPIENT_EMAIL '{recipient_email}' is not a valid email address")
            return False
        else:
            print(f"✅ RECIPIENT_EMAIL: {recipient_email}")
    else:
        print(f"ℹ️  RECIPIENT_EMAIL not set (will default to {gmail_address})")
    
    return True

def test_smtp_connection():
    """Test connection to Gmail SMTP server"""
    print("\n🔍 Testing Gmail SMTP connection...")
    
    try:
        import smtplib
        import ssl
        
        load_dotenv()
        gmail_address = os.getenv('GMAIL_ADDRESS')
        gmail_password = os.getenv('GMAIL_APP_PASSWORD')
        
        if not gmail_address or not gmail_password:
            print("❌ Gmail credentials not configured")
            return False
        
        # Test connection
        context = ssl.create_default_context()
        with smtplib.SMTP("smtp.gmail.com", 587) as server:
            server.starttls(context=context)
            server.login(gmail_address, gmail_password)
            print("✅ Gmail SMTP connection successful")
            return True
            
    except smtplib.SMTPAuthenticationError:
        print("❌ Gmail authentication failed")
        print("📖 Check your App Password and ensure 2FA is enabled")
        return False
    except Exception as e:
        print(f"❌ Gmail connection failed: {str(e)}")
        return False

def test_files():
    """Test if required files exist"""
    print("\n🔍 Testing required files...")
    
    required_files = [
        'contact-form.html',
        'app.py',
        'requirements.txt',
        '.env.example'
    ]
    
    all_exist = True
    for file in required_files:
        if os.path.exists(file):
            print(f"✅ {file} exists")
        else:
            print(f"❌ {file} missing")
            all_exist = False
    
    # Check .env file
    if os.path.exists('.env'):
        print("✅ .env exists")
    else:
        print("⚠️  .env file not found")
        print("📋 Copy .env.example to .env and configure your credentials")
        all_exist = False
    
    return all_exist

def main():
    """Run all tests"""
    print("🧪 Contact Form Setup Test")
    print("=" * 40)
    
    tests = [
        ("Dependencies", test_dependencies),
        ("Files", test_files),
        ("Environment", test_environment),
        ("Gmail Connection", test_smtp_connection)
    ]
    
    results = []
    for test_name, test_func in tests:
        try:
            result = test_func()
            results.append((test_name, result))
        except Exception as e:
            print(f"❌ {test_name} test failed with error: {str(e)}")
            results.append((test_name, False))
    
    # Summary
    print("\n" + "=" * 40)
    print("📊 Test Summary:")
    
    all_passed = True
    for test_name, passed in results:
        status = "✅ PASS" if passed else "❌ FAIL"
        print(f"  {status} {test_name}")
        if not passed:
            all_passed = False
    
    print("\n" + "=" * 40)
    if all_passed:
        print("🎉 All tests passed! Your contact form is ready to use.")
        print("🚀 Run 'python app.py' or './run.sh' to start the application")
    else:
        print("⚠️  Some tests failed. Please fix the issues above.")
        print("📖 Check SETUP_GUIDE.md for detailed instructions")
    
    return 0 if all_passed else 1

if __name__ == "__main__":
    sys.exit(main())